# Shopify Data Science Challenge

#### 100 sneaker stores sell the similar sneaker and given the transaction data it is required to verified the Average order value

#### Average Order Value (AOV) - the average of total orders made for the sneaker in all those stores 

$ AOV = (N1 Orders + N2 Orders + ..... + N100 Orders) / 5000 $

### inserting the relevant libraries

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

#### importing the data

In [4]:
sneakers = pd.read_csv("2019 Winter Data Science Intern Challenge Data Set - Sheet1.csv")
sneakers

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0            1       53      746           224            2           cash   
1            2       92      925            90            1           cash   
2            3       44      861           144            1           cash   
3            4       18      935           156            1    credit_card   
4            5       18      883           156            1    credit_card   
...        ...      ...      ...           ...          ...            ...   
4995      4996       73      993           330            2          debit   
4996      4997       48      789           234            2           cash   
4997      4998       56      867           351            3           cash   
4998      4999       60      825           354            2    credit_card   
4999      5000       44      734           288            2          debit   

               created_at  
0     2017-03-13 12:36:56  
1     2017-03-03 17:38:52  
2      2017-03-14 4:23:56  
3     2017-03-26 12:43:37  
4      2017-03-01 4:35:11  
...                   ...  
4995  2017-03-30 13:47:17  
4996  2017-03-16 20:36:16  
4997   2017-03-19 5:42:42  
4998  2017-03-16 14:51:18  
4999  2017-03-18 15:48:18  

[5000 rows x 7 columns]

#### there are visible difference between the total cost of orders on similar no of items sold which means each store is selling on different price point, let see...

In [18]:
sneakerCost = []

for order_id in sneakers:
    sneakerCost = sneakers["order_amount"] / sneakers["total_items"]

#sneakerCost
sneakers["Sneaker Cost"] = sneakerCost
sneakers = sneakers[["order_id", "shop_id", "user_id", "order_amount", "total_items", "Sneaker Cost", "payment_method",
                    "created_at"]]
sneakers

order_id  shop_id  user_id  order_amount  total_items  Sneaker Cost  \
0            1       53      746           224            2         112.0   
1            2       92      925            90            1          90.0   
2            3       44      861           144            1         144.0   
3            4       18      935           156            1         156.0   
4            5       18      883           156            1         156.0   
...        ...      ...      ...           ...          ...           ...   
4995      4996       73      993           330            2         165.0   
4996      4997       48      789           234            2         117.0   
4997      4998       56      867           351            3         117.0   
4998      4999       60      825           354            2         177.0   
4999      5000       44      734           288            2         144.0   

     payment_method           created_at  
0              cash  2017-03-13 12:36:56  
1              cash  2017-03-03 17:38:52  
2              cash   2017-03-14 4:23:56  
3       credit_card  2017-03-26 12:43:37  
4       credit_card   2017-03-01 4:35:11  
...             ...                  ...  
4995          debit  2017-03-30 13:47:17  
4996           cash  2017-03-16 20:36:16  
4997           cash   2017-03-19 5:42:42  
4998    credit_card  2017-03-16 14:51:18  
4999          debit  2017-03-18 15:48:18  

[5000 rows x 8 columns]

In [80]:
AOV = sneakers["order_amount"].sum()/5000
AOV

3145.128

#### below table verifies that the amount remains the same regardless of the payment method, as the max price and avg price are same.

In [59]:
shops = pd.DataFrame(columns={"shop_id", "Max Price", "Avg Price"})
shops["shop_id"] = sneakers["shop_id"].unique()
shops = shops.sort_values(by="shop_id")
shops.reset_index(inplace=True)
shopAvg = sneakers.groupby(by="shop_id").mean()
shopAvg.reset_index(inplace=True)
shopMax = sneakers.groupby(by="shop_id").max()
shopMax.reset_index(inplace=True)
shops = shops.drop(columns={"index"})
for i in range(0, len(shops)):
    shops["Avg Price"][i] = shopAvg["Sneaker Cost"][i]
    shops["Max Price"][i] = shopMax["Sneaker Cost"][i]

shops


C:\Users\Moiz\AppData\Roaming\Python\Python37\site-packages\ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # This is added back by InteractiveShellApp.init_path()
C:\Users\Moiz\AppData\Roaming\Python\Python37\site-packages\ipykernel_launcher.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  if sys.path[0] == '':


shop_id Max Price Avg Price
0         1       158       158
1         2        94        94
2         3       148       148
3         4       128       128
4         5       142       142
5         6       187       187
6         7       112       112
7         8       132       132
8         9       118       118
9        10       148       148
10       11       184       184
11       12       201       201
12       13       160       160
13       14       116       116
14       15       153       153
15       16       156       156
16       17       176       176
17       18       156       156
18       19       163       163
19       20       127       127
20       21       142       142
21       22       146       146
22       23       156       156
23       24       140       140
24       25       130       130
25       26       176       176
26       27       169       169
27       28       164       164
28       29       163       163
29       30       153       153
30       31       129       129
31       32       101       101
32       33       173       173
33       34       122       122
34       35       164       164
35       36       130       130
36       37       142       142
37       38       190       190
38       39       134       134
39       40       161       161
40       41       118       118
41       42       352       352
42       43       181       181
43       44       144       144
44       45       142       142
45       46       166       166
46       47       145       145
47       48       117       117
48       49       129       129
49       50       193       193
50       51       187       187
51       52       146       146
52       53       112       112
53       54       133       133
54       55       171       171
55       56       117       117
56       57       147       147
57       58       138       138
58       59       178       178
59       60       177       177
60       61       158       158
61       62       160       160
62       63       136       136
63       64       133       133
64       65       154       154
65       66       161       161
66       67       131       131
67       68       136       136
68       69       131       131
69       70       173       173
70       71       164       164
71       72       160       160
72       73       165       165
73       74       153       153
74       75       128       128
75       76       155       155
76       77       156       156
77       78     25725     25725
78       79       181       181
79       80       145       145
80       81       177       177
81       82       177       177
82       83       129       129
83       84       153       153
84       85       172       172
85       86       130       130
86       87       149       149
87       88       176       176
88       89       196       196
89       90       178       178
90       91       160       160
91       92        90        90
92       93       114       114
93       94       134       134
94       95       168       168
95       96       153       153
96       97       162       162
97       98       133       133
98       99       195       195
99      100       111       111

#### a) the value calculated originally is higher due to the price point set by store #78 of 25725 (a possible outlier). Which eventually is moving the average way up.

#### b) we could instead use modal value (mode), that will tell the consistent amount a consumer is paying throughout all these stores. 

In [81]:
sneakers["order_amount"].mode()


0    153
dtype: int64

In [82]:
shops["Max Price"].mode()

0    153
dtype: object

#### c) here above the modal order amount is 153 is what seems reasonable as many orders have been made out around this amount and interestingly it is price that most stores have set for the sneakers!

## SQL Queries

#### 1) SELECT count(o.OrderID) from Orders o LEFT JOIN Shippers s ON o.ShipperID = s.ShipperID WHERE s.ShipperName = "Speedy Express";

####  54

#### 2) SELECT e.LastName, count(*) as EmpOrdrs, o.EmployeeID FROM Employees e JOIN Orders o ON e.EmployeeID = o.EmployeeID GROUP BY e.EmployeeID ORDER BY count(*) desc LIMIT 1;

#### Peacock 40 4

#### 3) SELECT od.ProductID, od.Quantity, c.Country, p.ProductName FROM OrderDetails od LEFT JOIN Orders o ON od.OrderID = o.OrderID JOIN Customers c ON o.CustomerID = c.CustomerID JOIN Products p ON od.ProductID = p.ProductID WHERE c.Country = "Germany" GROUP BY od.ProductID ORDER BY od.Quantity desc LIMIT 1;

#### 35	100	Germany	Steeleye Stout